src/sqlfluff/rules/L064.py

```diff
@@ -0,0 +1,259 @@
+                    )
+                ],
+                description=(
+                    "Inconsistent use of preferred quote style '"
+                    f"{self._quotes_mapping[preferred_string_quotes]['common_name']}'. "
+                    f"Use {fixed_string} instead of {context.segment.raw}."
+                ),
+            )
+
+        return None
+
+    def _sub_twice(self, regex: regex.Pattern, replacement: str, original: str) -> str:
+        """Replace `regex` with `replacement` twice on `original`.
+
+        This is used by string normalization to perform replaces on overlapping matches.
+        """
+        return regex.sub(replacement, regex.sub(replacement, original))
+
+    def _normalize_preferred_string_quotes(
+        self, s: str, preferred_quote_char: str, alternate_quote_char: str
+    ) -> str:
+        """Prefer `preferred_quote_char` but only if it doesn't cause more escaping.
+
+        Adds or removes backslashes as appropriate.
+        """
+        value = s.lstrip(self._string_prefix_chars)
+
+        if value[:3] == preferred_quote_char * 3:
+            # In tripple quoted strings we are not replacing escaped quotes.
+            # So nothing left to do and we can exit.
+            return s
+        elif value[0] == preferred_quote_char:
+            # Quotes are alright already. But maybe we can remove some unnecessary
+            # escapes or reduce the number of escapes using alternate_quote_char ?
+            orig_quote = preferred_quote_char
+            new_quote = alternate_quote_char
+        elif value[:3] == alternate_quote_char * 3:
+            orig_quote = alternate_quote_char * 3
+            new_quote = preferred_quote_char * 3
+        elif value[0] == alternate_quote_char:
+            orig_quote = alternate_quote_char
+            new_quote = preferred_quote_char
+        else:
+            self.logger.debug(
+                "Found quoted string %s using neither preferred quote char %s "
+                "nor alternate_quote_char %s. Skipping...",
+                s,
+                preferred_quote_char,
+                alternate_quote_char,
+            )
+            return s
+
+        first_quote_pos = s.find(orig_quote)
+        prefix = s[:first_quote_pos]
+        unescaped_new_quote = regex.compile(rf"(([^\\]|^)(\\\\)*){new_quote}")
+        escaped_new_quote = regex.compile(rf"([^\\]|^)\\((?:\\\\)*){new_quote}")
+        escaped_orig_quote = regex.compile(rf"([^\\]|^)\\((?:\\\\)*){orig_quote}")
+        body = s[first_quote_pos + len(orig_quote) : -len(orig_quote)]
+
+        if "r" in prefix.lower():
+            if unescaped_new_quote.search(body):
+                self.logger.debug(
+                    "There's at least one unescaped new_quote in this raw string "
+                    "so converting is impossible."
+                )
+                return s
+
+            # Do not introduce or remove backslashes in raw strings
+            new_body = body
```

Comment:
This comment confuses me -- do we **know** that we have introduced or removed backslashes here? Would it be correct (and clearer) to say "Do not modify the body of raw strings"?

Reply:
Hmm, I guess it's a bit of personal taste question. To me the from the code itself it's quite clear that the body of raw strings is not getting modified, or at least `new_body = body` would suggest that to me. The question that needs to be answered is: Why are bodies of raw strings not modified? Because we don't want to introduce or remove escaped quotes, as this changes the actual string value.

I will try to make this more clear.